File: FastaFileSplitterDLL/clsFastaOutputFile.vb
Option Strict On

Imports System.IO

Public Class clsFastaOutputFile

#Region "Constants and Enums"
    Public Const DEFAULT_PROTEIN_LINE_START_CHAR As Char = ">"c
    Public Const DEFAULT_PROTEIN_LINE_ACCESSION_END_CHAR As Char = " "c
    Public Const DEFAULT_RESIDUES_PER_LINE As Integer = 60
#End Region

#Region "Classwide Variables"

    Protected mOutputFileIsOpen As Boolean
    Protected mOutputFilePath As String
    Protected mOutputFile As StreamWriter

    Protected mProteinLineStartChar As String
    Protected mProteinLineAccessionEndChar As String

    Protected mResiduesPerLine As Integer

    Protected mTotalProteinsInFile As Integer
    Protected mTotalResiduesInFile As Int64

#End Region

#Region "Properties"

    Public ReadOnly Property OutputFileIsOpen As Boolean
        Get
            Return mOutputFileIsOpen
        End Get
    End Property

    Public ReadOnly Property OutputFilePath As String
        Get
            Return mOutputFilePath
        End Get
    End Property

    Public Property ResiduesPerLine As Integer
        Get
            Return mResiduesPerLine
        End Get
        Set
            If Value < 1 Then Value = 1
            mResiduesPerLine = Value
        End Set
    End Property

    Public ReadOnly Property TotalProteinsInFile As Integer
        Get
            Return mTotalProteinsInFile
        End Get
    End Property

    Public ReadOnly Property TotalResiduesInFile As Int64
        Get
            Return mTotalResiduesInFile
        End Get
    End Property
#End Region

    Public Sub New(outputFilePath As String)
        Me.New(outputFilePath, DEFAULT_PROTEIN_LINE_START_CHAR, DEFAULT_PROTEIN_LINE_ACCESSION_END_CHAR)
    End Sub

    Public Sub New(outputFilePath As String, proteinLineStartChar As Char, proteinLineAccessionEndChar As Char)
        If outputFilePath Is Nothing OrElse outputFilePath.Length = 0 Then
            Throw New Exception("OutputFilePath is empty; cannot instantiate class")
        End If
        mOutputFilePath = outputFilePath

        mOutputFile = New StreamWriter(New FileStream(outputFilePath, FileMode.Create, FileAccess.Write, FileShare.Read))
        mOutputFileIsOpen = True

        mProteinLineStartChar = proteinLineStartChar
        mProteinLineAccessionEndChar = proteinLineAccessionEndChar

        mResiduesPerLine = DEFAULT_RESIDUES_PER_LINE

        mTotalProteinsInFile = 0
        mTotalResiduesInFile = 0
    End Sub

    Public Sub CloseFile()
        Try
            If mOutputFileIsOpen AndAlso Not mOutputFile Is Nothing Then
                mOutputFile.Close()
                mOutputFileIsOpen = False
            End If
        Catch ex As Exception
            ' Ignore errors here
        End Try
    End Sub

    Public Sub StoreProtein(proteinName As String, ByRef description As String, ByRef sequence As String)

        If mOutputFileIsOpen Then

            Try
                ' Write out the protein header and description line
                mOutputFile.WriteLine(mProteinLineStartChar & proteinName & mProteinLineAccessionEndChar & description)

                ' Now write out the residues, storing mResiduesPerLine residues per line
                Dim startIndex = 0
                Do While startIndex < sequence.Length
                    Dim charCount = mResiduesPerLine
                    If startIndex + charCount > sequence.Length Then
                        charCount = sequence.Length - startIndex
                    End If
                    mOutputFile.WriteLine(sequence.Substring(startIndex, charCount))
                    startIndex += charCount
                Loop

                mTotalProteinsInFile += 1
                mTotalResiduesInFile += sequence.Length

            Catch ex As Exception
                Console.WriteLine("Error in StoreProtein: " & ex.Message)
                Throw
            End Try

        End If

    End Sub

End Class

File: FastaFileSplitterDLL/clsFastaOutputFile.cs
﻿using System;
using System.IO;
using PRISM;

namespace FastaFileSplitterLibrary
{
    /// <summary>
    /// This class is used to create the split FASTA files
    /// </summary>
    public class clsFastaOutputFile : EventNotifier
    {
        /// <summary>
        /// Protein line start character
        /// </summary>
        public const char DEFAULT_PROTEIN_LINE_START_CHAR = '>';

        /// <summary>
        /// Protein name / description separator character
        /// </summary>
        public const char DEFAULT_PROTEIN_LINE_ACCESSION_END_CHAR = ' ';

        /// <summary>
        /// Number of residues per line for the protein sequence
        /// </summary>
        public const int DEFAULT_RESIDUES_PER_LINE = 60;

        private readonly StreamWriter mOutputFile;
        private readonly string mProteinLineStartChar;
        private readonly string mProteinLineAccessionEndChar;
        private int mResiduesPerLine;
        private int mTotalProteinsInFile;
        private long mTotalResiduesInFile;

        /// <summary>
        /// True if the output file is open for writing
        /// </summary>
        public bool OutputFileIsOpen { get; private set; }

        /// <summary>
        /// Output file path
        /// </summary>
        public string OutputFilePath { get; }

        /// <summary>
        /// Residues per line to write to the output file
        /// </summary>
        // ReSharper disable once UnusedMember.Global
        public int ResiduesPerLine
        {
            get => mResiduesPerLine;
            set
            {
                if (value < 1)
                    value = 1;
                mResiduesPerLine = value;
            }
        }

        /// <summary>
        /// Total proteins in the file
        /// </summary>
        public int TotalProteinsInFile => mTotalProteinsInFile;

        /// <summary>
        /// Total residues in the file
        /// </summary>
        public long TotalResiduesInFile => mTotalResiduesInFile;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="outputFilePath"></param>
        /// <param name="proteinLineStartChar"></param>
        /// <param name="proteinLineAccessionEndChar"></param>
        public clsFastaOutputFile(
            string outputFilePath,
            char proteinLineStartChar = DEFAULT_PROTEIN_LINE_START_CHAR,
            char proteinLineAccessionEndChar = DEFAULT_PROTEIN_LINE_ACCESSION_END_CHAR)
        {
            if (string.IsNullOrWhiteSpace(outputFilePath))
            {
                throw new Exception("OutputFilePath is empty; cannot instantiate class");
            }

            OutputFilePath = outputFilePath;

            mOutputFile = new StreamWriter(new FileStream(outputFilePath, FileMode.Create, FileAccess.Write, FileShare.Read));

            OutputFileIsOpen = true;

            mProteinLineStartChar = proteinLineStartChar.ToString();
            mProteinLineAccessionEndChar = proteinLineAccessionEndChar.ToString();

            mResiduesPerLine = DEFAULT_RESIDUES_PER_LINE;

            mTotalProteinsInFile = 0;
            mTotalResiduesInFile = 0L;
        }

        /// <summary>
        /// Close the output file
        /// </summary>
        public void CloseFile()
        {
            try
            {
                if (!OutputFileIsOpen)
                {
                    return;
                }

                mOutputFile.Close();
                OutputFileIsOpen = false;
            }
            catch (Exception ex)
            {
                OnWarningEvent("Error closing the output file: " + ex.Message);
            }
        }

        /// <summary>
        /// Append a protein to the output file
        /// </summary>
        /// <param name="proteinName"></param>
        /// <param name="description"></param>
        /// <param name="sequence"></param>
        public void StoreProtein(string proteinName, string description, string sequence)
        {
            if (!OutputFileIsOpen)
            {
                return;
            }

            try
            {
                // Write out the protein header and description line
                mOutputFile.WriteLine(mProteinLineStartChar + proteinName + mProteinLineAccessionEndChar + description);

                // Now write out the residues, storing mResiduesPerLine residues per line
                var startIndex = 0;
                while (startIndex < sequence.Length)
                {
                    var charCount = mResiduesPerLine;
                    if (startIndex + charCount > sequence.Length)
                    {
                        charCount = sequence.Length - startIndex;
                    }

                    mOutputFile.WriteLine(sequence.Substring(startIndex, charCount));
                    startIndex += charCount;
                }

                mTotalProteinsInFile++;
                mTotalResiduesInFile += sequence.Length;
            }
            catch (Exception ex)
            {
                OnErrorEvent("Error in StoreProtein", ex);
                throw;
            }
        }
    }
}